EIPS/eip-7923.md

```diff
@@ -22,11 +22,19 @@ The EVM currently uses a quadratic pricing model for its memory. This was origin
 2. The quadratic model makes it difficult to reason about how much memory a transaction can allocate. It requires solving an optimization problem which involves computing how many message calls are available to recurse into based on the call stack limit (and, post [EIP-150](./eip-150.md), the 63/64ths rule), and then maximizing the memory used per message call.
 3. The quadratic model makes it impossible for high-level smart contracts languages to get the benefits of virtual memory. Most modern programming languages maintain what is known as the "heap" and the "call stack". The heap is used to allocate objects which live past the lifetime of their current function frame, whereas the call stack is used to allocate objects which live in the current function frames. Importantly, the call stack starts at the top of memory and grows down, while the heap starts at the bottom of memory and grows up, thus the language implementation does not need to worry about the two regions of memory interfering with each other. This is a feature which is enabled by virtual, paged memory, which has been present in operating systems since the early 90's. However, smart contract languages like Vyper and Solidity are not able to implement this, leading to inefficiencies in their memory models.
 
-This EIP proposes a linear costing model which more closely reflects the hardware of today. It uses a virtual addressing scheme so that memory pages are not allocated until they are actually accessed. Notably, the data structures used for costing memory do not need to be part of the memory implementation itself, which suggests an elegant implementation using `mmap`.
+This EIP proposes a linear costing model which more closely reflects the hardware of today, which is hierarchical ("hot" memory is much faster to access than "cold" memory), and virtually addressed (memory does not need to be allocated contiguously, but is rather served "on-demand" by the operating system).
+
+First, some preliminaries. A page is 4096 bytes on most architectures. Given a memory address, its page is simple to compute by masking out the rightmost 12 bits.
+
+There are two factors which contribute to "cold" memory (i.e. not-recently-used) being slower: CPU cache and TLB (Translation Lookaside Buffer) cache. The CPU cache is a least-recently-used memory cache, which is significantly faster than fetching all the way from RAM. The TLB is usually some hash table which maps virtual pages (used by the user) to physical pages in RAM. "Thrashing", or accessing a lot of different memory addresses, does two things: it pushes memory out of the hot cache and into cold memory, and it pushes pages out of the TLB cache.
```

Comment:
Ah ok, this note on thrashing is great. The gas cost for thrashing is thus not for thrashing the page we are currently creating/updating, it is rather that we thrash another one (which goes hot -> cold). (Do we want to charge the thrash gas if the LRU size is lower than the max? Since in this, items in the LRU are "hot" and items not in the LRU are "cold", but if we push the first page to the LRU then we thus also do not move another page to the cold memory)

Reply:
This is actually quite a complex topic (cache coherence - https://en.wikipedia.org/wiki/Cache_coherence), and when exactly a byte comes into and falls out of cache might be fairly complicated, depending on heuristics, threading and other factors. But in general I think we want to think of the cost as when the page is pulled into cache. Kicking pages out of the cache is relevant insofar as it makes those pages expensive to fetch later on -- if they are never kicked out of the cache, we could just charge 100 gas at allocation and call it a day (since later accesses to those pages would be hot).